--- selectsource.php.orig	2019-11-12 12:01:55.000000000 +0100
+++ selectsource.php	2019-11-12 12:50:01.000000000 +0100
@@ -1,9 +1,11 @@
 <?php
 
+use SimpleSAML\SessionHandler;
+
 /**
  * This page shows a list of authentication sources. When the user selects
  * one of them if pass this information to the
- * \SimpleSAML\Module\multiauth\Auth\Source\MultiAuth class and call the
+ * \SimpleSAML\Module\sgis\Auth\Source\MultiAuth class and call the
  * delegateAuthentication method on it.
  *
  * @author Lorenzo Gil, Yaco Sistemas S.L.
@@ -15,7 +17,7 @@ if (!array_key_exists('AuthState', $_REQ
     throw new \SimpleSAML\Error\BadRequest('Missing AuthState parameter.');
 }
 $authStateId = $_REQUEST['AuthState'];
-$state = \SimpleSAML\Auth\State::loadState($authStateId, \SimpleSAML\Module\multiauth\Auth\Source\MultiAuth::STAGEID);
+$state = \SimpleSAML\Auth\State::loadState($authStateId, \SimpleSAML\Module\sgis\Auth\Source\MultiAuth::STAGEID);
 
 if (array_key_exists("\SimpleSAML\Auth\Source.id", $state)) {
     $authId = $state["\SimpleSAML\Auth\Source.id"];
@@ -35,25 +37,42 @@ if (array_key_exists('source', $_REQUEST
         }
     }
 }
+if ($as !== null && $as->getRememberSourceEnabled() && $source === NULL && array_key_exists($as->getAuthId() . '-source', $_COOKIE)) {
+    $source = $_COOKIE[$as->getAuthId() . '-source'];
+    $as->setPreviousSource($source);
+    \SimpleSAML\Module\sgis\Auth\Source\MultiAuth::delegateAuthentication($source, $state);
+}
+if ($source !== NULL && $as->getRememberSourceEnabled()) {
+    $sessionHandler = SessionHandler::getSessionHandler();
+    $params = $sessionHandler->getCookieParams();
+    $params['expire'] = time();
+    $params['expire'] += (isset($_REQUEST['remember_source']) && $_REQUEST['remember_source'] == 'Yes' ? 31536000 : -300);
+    setcookie($as->getAuthId() . '-source', $source, $params['expire'], $params['path'], $params['domain'], $params['secure'], $params['httponly']);
+}
 if ($source !== null) {
     if ($as !== null) {
         $as->setPreviousSource($source);
     }
-    \SimpleSAML\Module\multiauth\Auth\Source\MultiAuth::delegateAuthentication($source, $state);
+    \SimpleSAML\Module\sgis\Auth\Source\MultiAuth::delegateAuthentication($source, $state);
 }
 
 if (array_key_exists('multiauth:preselect', $state)) {
     $source = $state['multiauth:preselect'];
-    \SimpleSAML\Module\multiauth\Auth\Source\MultiAuth::delegateAuthentication($source, $state);
+    \SimpleSAML\Module\sgis\Auth\Source\MultiAuth::delegateAuthentication($source, $state);
 }
 
 $globalConfig = \SimpleSAML\Configuration::getInstance();
-$t = new \SimpleSAML\XHTML\Template($globalConfig, 'multiauth:selectsource.php');
+$t = new \SimpleSAML\XHTML\Template($globalConfig, 'sgis:selectsource.php');
 
 $defaultLanguage = $globalConfig->getString('language.default', 'en');
 $language = $t->getTranslator()->getLanguage()->getLanguage();
 
-$sources = $state[\SimpleSAML\Module\multiauth\Auth\Source\MultiAuth::SOURCESID];
+$sources = $state[\SimpleSAML\Module\sgis\Auth\Source\MultiAuth::SOURCESID];
+if (count($sources) == 1) {
+    $sourceKey = array_shift(array_keys($sources));
+    $source = $sources[$sourceKey]['source'];
+    \SimpleSAML\Module\sgis\Auth\Source\MultiAuth::delegateAuthentication($source, $state);
+}
 foreach ($sources as $key => $source) {
     $sources[$key]['source64'] = base64_encode($sources[$key]['source']);
     if (isset($sources[$key]['text'][$language])) {
@@ -78,5 +97,7 @@ if ($as !== null) {
 } else {
     $t->data['preferred'] = null;
 }
+$t->data['rememberSourceEnabled'] = ($as !== null && $as->getRememberSourceEnabled());
+$t->data['rememberSourceChecked'] = ($as !== null && $as->getRememberSourceChecked());
 $t->show();
 exit();
