--- selectsource.php.orig	2019-11-12 11:57:28.000000000 +0100
+++ selectsource.php	2019-11-12 11:57:04.000000000 +0100
@@ -21,10 +21,18 @@ foreach ($this->data['sources'] as $sour
     $name = 'src-'.base64_encode($source['source']);
     echo '<input type="submit" name="'.htmlspecialchars($name).'"'.$autofocus.' '.
         'id="button-'.htmlspecialchars($source['source']).'" '.
-        'value="'.htmlspecialchars($this->t($source['text'])).'" />';
+        'value="'.htmlspecialchars($source['text']).'" />';
     echo '</li>';
 }
 ?>
     </ul>
+<?php
+if ($this->data['rememberSourceEnabled']) {
+   	echo str_repeat("\t", 4);
+    echo '<input type="checkbox" id="remember_source" tabindex="4" name="remember_source" value="Yes" ';
+    echo ($this->data['rememberSourceChecked'] ? 'checked="Yes" /> ' : '/> ');
+    echo $this->t('{sgis:sgis:remember_source}');
+}
+?>
 </form>
 <?php $this->includeAtTemplateBase('includes/footer.php');
